src/stamps/extends.ts

```diff
@@ -0,0 +1,163 @@
+      .sort((a, b) => (getUsage(a) < getUsage(b) ? 1 : -1))
+
+    // return the all usable stamp sorted by usage
+    return usableStamps
+  }
+
+  return []
+}
+
+export async function topUpStamp(beeDebug: BeeDebug, postageBatchId: string, amount: string): Promise<PostageBatch> {
+  await beeDebug.topUpBatch(postageBatchId, amount)
+  const stamp = await beeDebug.getPostageBatch(postageBatchId)
+
+  return stamp
+}
+
+export class ExtendsStampManager extends BaseStampManager {
+  private isBuyingStamp?: boolean = false
+  private topingUpStamps: string[] = []
+  /**
+   * Start the manager in either hardcoded or autobuy mode
+   */
+  async start(config: StampsConfigExtends): Promise<void> {
+    // Extends mode
+    const refreshStamps = async () => this.refreshStampsExtends(config, new BeeDebug(config.beeDebugApiUrl))
+
+    this.stop()
+    await refreshStamps()
+
+    this.interval = setInterval(refreshStamps, config.refreshPeriod)
+  }
+
+  completeTopUp(extendsTypeFeature: 'ttl' | 'capacity', stamp: PostageBatch) {
+    if (extendsTypeFeature === 'ttl') {
+      logger.info('successfully postage stamp TTL extended', { stamp })
+    } else {
+      logger.info('successfully postage stamp capacity extended', { stamp })
+    }
+    // remove stamps from extending stamps array
+    const stampIndex = this.topingUpStamps.findIndex(id => stamp.batchID === id)
+    this.topingUpStamps.splice(stampIndex, 1)
+  }
+
+  async verifyUsableStamps(beeDebug: BeeDebug, ttlMin: number, config: StampsConfigExtends, amount: string) {
+    for (let i = 0; i < this.usableStamps!.length; i++) {
+      const stamp = this.usableStamps![i]
+
+      const minTimeThreshold = ttlMin + config.refreshPeriod / 1000
+
+      if (
+        ttlMin > 0 &&
+        stamp.batchTTL < minTimeThreshold &&
+        !this.topingUpStamps.includes(stamp.batchID) &&
+        amount !== '0'
+      ) {
+        this.topingUpStamps.push(stamp.batchID)
+        logger.info(`extending postage stamp TTL ${stamp.batchID}`)
+
+        try {
+          const stampRes = await topUpStamp(beeDebug, stamp.batchID, amount)
+
+          setTimeout(() => this.completeTopUp('ttl', stampRes), 60000)
+        } catch (e: any) {
+          // error that indicate that 2 stamps are trying to be extended at the same time. Comes out as a warning
+          const errorStampIndex = this.topingUpStamps.indexOf(stamp.batchID)
+          this.topingUpStamps.splice(errorStampIndex, 1)
+          logger.error('failed to topup postage stamp', e)
+        }
+      }
+    }
+  }
+
+  public async refreshStampsExtends(config: StampsConfigExtends, beeDebug: BeeDebug): Promise<void> {
+    let usableStampsExtendsTTL: PostageBatch[] = []
+    let usableStampsExtendsCapacity: PostageBatch[] = []
+    stampCheckCounter.inc()
+    logger.info('checking postage stamps')
+
+    try {
+      const stamps = await beeDebug.getAllPostageBatch()
+
+      const { depth, amount, ttlMin, usageThreshold } = config
```

Comment:
I would expect to have some flags that would enable/disable extending TTL and Capacity... How do you distinguish that one should work and the other not?

Reply:
It runs both at the same time, depending if the filter methods return any stamp at all according to the conditions. For instance, to filter TTL checks the `batchTTL`of the stamps and to filter capacity matches the stamp's usage